src/purifier/purifier.js

```diff
@@ -65,13 +65,12 @@ export let AttributeRewriterDef;
 
 export class Purifier {
   /**
-   * @param {!Document} doc
    * @param {!JsonObject=} opt_config
    * @param {!AttributeRewriterDef=} opt_attrRewrite
    */
-  constructor(doc, opt_config, opt_attrRewrite) {
-    /** @private {!Document} */
-    this.doc_ = doc;
+  constructor(opt_config, opt_attrRewrite) {
+    /** @private {Document} */
+    this.doc_ = null;
```

Comment:
Why move this out of the constructor and into each function?

Reply:
How else could it be a window level service that could operate on all the ampdocs of the page unless it were a window service?

Reply:
definitely the messiest part of this PR, since the purifier hooks are registered in the constructor but the doc it operates on changes each invocation

Reply:
Hm, does `this.doc_` actually change across invocations? 

A bit confusing but there can be multiple AmpDoc per window but only one Document. I.e. PWAs have one AmpDoc per shadow AMP but share a single Document. AMP ads have their own window since they're iframed.

Reply:
I'm so glad for this question then :). I can basically revert all of the changes to `purifier.js` now. I had assumed they had different `document`s.